metagraph/src/graph/representation/succinct/dbg_succinct_cached.hpp

```diff
@@ -0,0 +1,152 @@
+#ifndef __DBG_SUCCINCT_CACHED__
+#define __DBG_SUCCINCT_CACHED__
+
+#include "dbg_succinct.hpp"
+#include "kmer/kmer_extractor.hpp"
+
+namespace mtg {
+namespace graph {
+
+
+template <typename KmerType>
+class DBGSuccinctCachedViewImpl : public DBGSuccinct::CachedView {
+  public:
+    template <typename Graph>
+    DBGSuccinctCachedViewImpl(Graph&& graph, size_t cache_size)
+          : DBGSuccinct::CachedView(std::forward<Graph>(graph)),
+            boss_(&graph_->get_boss()),
+            decoded_cache_(cache_size) {}
+
+    virtual TAlphabet get_first_value(edge_index i) const override final {
+        assert(i);
+
+        // Take k - 1 traversal steps to construct the node sequence. This way,
+        // the last node visited and its parent can be cached for subsequence calls.
+        auto kmer_pair = get_kmer_pair(i, true);
+        assert(kmer_pair.second);
+
+        TAlphabet c = kmer_pair.first[1];
+
+        edge_index bwd_i = boss_->bwd(i);
+        auto fetch_bwd = decoded_cache_.TryGet(bwd_i);
+        if (!fetch_bwd) {
+            kmer_pair.second = boss_->bwd(*kmer_pair.second);
+            kmer_pair.first.to_prev(get_k(), boss_->get_node_last_value(*kmer_pair.second));
+            put_kmer(bwd_i, std::move(kmer_pair));
+        }
```

Comment:
An alternative:
instead of
```
i <- bwd(i) <- ... (k bwd steps) ... <- bwd(...(...))
^    *                                  ^
```
cache
```
i <- parent <- ... (arbitrary back steps) ... <- k-order-parent
^    ^                                           ^
```
and now you can cache arbitrary paths, not necessarily bwd ... as before